Make gp_dump_query_oids an internal function, and fix compiler warning & segfault.

hlinnaka opened the pull request:
No need to create a whole new library for this. It's included in every
installation anyway, as part of gp_toolkit. Might as well just include
it in the backend, and make the build slightly simpler.

hlinnaka commented:
Note: gp_dump_query_oids currently segfaults if you pass it an EXPLAIN query. There's a related compiler warning about that too. I started looking into that, but decided to clean up the build system first.

hlinnaka commented:
@hsyuan, digging into the git history, it seems that I already fixed the compiler warning once:

```
commit 754d0cc46984c92e8d465c8e2287a268a47bc01e
Author: Heikki Linnakangas <hlinnakangas@pivotal.io>
Date:   Sat Jun 18 23:08:58 2016 +0300

    Fix crash if an EXPLAIN statement is passed to gp_dump_query_oids()
    
    I was alerted to this by a compiler warning, which this also fixes. Was
    broken by the 8.3 merge.
```

But you reverted that change with:
```
commit 974c13d4f32787c9edd71e84e983d3d81fa6ec92
Author: Haisheng Yuan and Marc Spehlmann <marc.spehlmann@gmail.com>
Date:   Mon Aug 8 14:33:38 2016 -0700

    [#127944191] Fix compiler warning in gp_dump_query_oids
    
    We just need to pass the query tree, we don't need source sql text and
    other arguments, so change to QueryRewrite here. Previously an implicit
    cast was causing a compiler warning. Also, the method pg_analyze_and_rewrite
    is overkill, and in fact calls QueryRewrite.
```

What happened there?

This PR fixes that again, as well as cleans it up in general and adds a regression test, but what was the reason for reverting it previously? How does this PR look to you, aside from fixing/reverting the warning?

hsyuan commented:
Previously we saw an implicit cast warning and compile error in SUSE because `ExplainStmt->query` is `Query*` instead of `Node*` in PG 8.3.

hsyuan commented on src/backend/utils/adt/gp_dump_oids.c:
```diff
@@ -83,50 +75,60 @@ traverseQueryOids
+	/*
+	 * Traverse through the query list. For EXPLAIN statements, the query list contains
+	 * an ExplainStmt with the raw parse tree of the actual query. Analyze the explained
+	 * queries instead of the ExplainStmt iteself.
```
typo. iteself->itself

hsyuan commented:
LGTM. I like your changes to just include it in the backend.
BTW, is the explain stmt crash caused by the merge of 8.3? Previously we tested statement with explain and it worked before.

hlinnaka commented:
The merge broke it. I fixed that in commit 754d0cc4. But then 974c13d4 reverted that change, and broke it again.

Fixed the typo and pushed. Thanks!
